### ISSUE TITLE
Memory leak with tf.py_function in eager mode using TF 2.0

### ISSUE THREAD

**Issue by workingloong** [tensorflow/tensorflow#35010]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.14.6
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0
- Python version: 3.6.3 64bit
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
I customize a Keras layer and use `tf.py_function` in the `call` of the layer. The used memory keeps increasing linearly with the number of calling the layer with inputs in eager mode. What's 

![image](https://user-images.githubusercontent.com/18071380/70585663-d5744a00-1bff-11ea-8d41-3649d035e9be.png)

**Describe the expected behavior**
The memory should keep stable.

**Code to reproduce the issue**
```python
import psutil
import numpy as np
import tensorflow as tf


class TestLayer(tf.keras.layers.Layer):
    def __init__(
        self,
        output_dim,
        **kwargs
    ):
        super(TestLayer, self).__init__(**kwargs)
        self.output_dim = output_dim

    def build(self, input_shape):
        self.built = True

    def call(self, inputs):
        batch_embedding = tf.py_function(
            self.mock_output, inp=[inputs], Tout=tf.float64,
        )
        return batch_embedding

    def mock_output(self, inputs):
        shape = inputs.shape.as_list()
        batch_size = shape[0]
        return tf.constant(np.random.random((batch_size,self.output_dim)))


test_layer = TestLayer(1000)

for i in range(1000):
    test_layer.call(tf.constant(np.random.randint(0,100,(256,10))))
    if i % 100 == 0:
        used_mem = psutil.virtual_memory().used
        print('used memory: {} Mb'.format(used_mem / 1024 / 1024))
```

**other info**
We encounter this issue when developing a customized embedding layer in [ElasticDL](https://github.com/sql-machine-learning/elasticdl) and resolving the ElasticDL [issue 1567](https://github.com/sql-machine-learning/elasticdl/issues/1567)

**Comment by skydoorkai** [comment]
Found some issues in [`_internal_py_func`](https://github.com/tensorflow/tensorflow/blob/2a2c812ab2330c9aac33335f10679a346436acfb/tensorflow/python/ops/script_ops.py#L273) in script_ops.py

Every time the py_function is called in the eager mode, a new `EagerFunc` is [created](https://github.com/tensorflow/tensorflow/blob/2a2c812ab2330c9aac33335f10679a346436acfb/tensorflow/python/ops/script_ops.py#L295), and a new `token` is[ added into](https://github.com/tensorflow/tensorflow/blob/2a2c812ab2330c9aac33335f10679a346436acfb/tensorflow/python/ops/script_ops.py#L311) `_py_funcs`. 
Also, this `func` is[ appended into](https://github.com/tensorflow/tensorflow/blob/2a2c812ab2330c9aac33335f10679a346436acfb/tensorflow/python/ops/script_ops.py#L333) `graph._py_funcs_used_in_graph`.

Looks like this logic is for graph mode, and has a side effect in the eager mode.

`_py_funcs` and  `graph._py_funcs_used_in_graph` will continously increasing in the eager mode.

This only accounts for a very small portion of the memory leak in the above sample code. 
There must be other reasons for the huge memory leak.

**Comment by ravikyram** [comment]
@workingloong 
I have tried in colab with TF 2.0, 2.1.0-rc0. With eager execution i can see huge memory leak ([gist attached](https://colab.sandbox.google.com/gist/ravikyram/c2bba090501f9a1aab13e1e86d2aea15/untitled466.ipynb)).By disabling eager execution i am seeing very small portion of memory leak ([gist attached)](https://colab.sandbox.google.com/gist/ravikyram/451e3366ffda46c971b3b54f169d22d2/untitled467.ipynb).Please, confirm is this the expected behavior?.Thanks!

**Comment by workingloong** [comment]
> @workingloong
> I have tried in colab with TF 2.0, 2.1.0-rc0. With eager execution i can see huge memory leak ([gist attached](https://colab.sandbox.google.com/gist/ravikyram/c2bba090501f9a1aab13e1e86d2aea15/untitled466.ipynb)).By disabling eager execution i am seeing very small portion of memory leak ([gist attached)](https://colab.sandbox.google.com/gist/ravikyram/451e3366ffda46c971b3b54f169d22d2/untitled467.ipynb).Please, confirm is this the expected behavior?.Thanks!

Yes, there is a very portion of memory leak in graph mode by using `@tf.function` to decorate the `call` in the customized layer.  However, i can not convert the inputs to a numpy array and use python to process the array because the inputs is a Tensor not an EagerTensor in graph mode. So i hope i can use the `tf.py_function`with eager execution.

```python
@tf.function
def call(self, inputs):
    batch_embedding = tf.py_function(
        self.mock_output, inp=[inputs], Tout=tf.float64,
    )
    return batch_embedding
```

**Comment by loretoparisi** [comment]
@skydoorkai we have found a similar issue when using

```python
results = tf.py_function(
                    func=self.safe_load,
                    inp=[audio_descriptor, offset, duration, sample_rate, dtype],
                    Tout=(tf.float32, tf.bool)),
                waveform, error = results[0]
```

where `self._safe_load` loads a wave file. This caused a leak of ~24MB (the wave file size) at every prediction, afterwards calling

```python
tf.reset_default_graph()
tf.keras.backend.clear_session()
```

we only see a leak of ~600-700KB at every call.

**Comment by TheCrazyT** [comment]
I had a similar problem on kaggle.
TLDR: try tf_nightly-2.3.0.dev20200520 (or newer)

I'm using TFRecordDataset and run_eagerly on compile (sadly I am currently forced to use eargerly because of another problem that happens in a custom loss function).
The memory usage gets increased after every epoch, so i can not execute the amount of epochs i want because the kernel dies because of "out of memory".
(Raising slowly from 6 GB to over 16 GB with each epoch)
I am not shure if it is a problem about tensorflow or keras.

Garbage collection, clear_session, event setting every variable to None etc. did not help.
Somehow the memory was still used.

Somehow it seems to be better with tf_nightly-2.3.0.dev20200520.
So maybe the bug was fixed?

**Comment by juanma9613** [comment]
I'm experiencing a memory leakage when training Mobilenetv2,Resnet, etc on Imagenet from scratch using a tf.py_function to map my dataset with a custom augmentation. It's not a huge leakage, but after three epochs of training on Imagenet dataset, my memory consumption increased from 16Gb to 59Gb and I had to stop my execution, because it started to be too slow.

I'm using tensorflow 2.3 on a p3.2xlarge instance for benchmarking purposes, I hope someone has a good way to deal with this, thank you all for your comments.

**Comment by mayou36** [comment]
Hi all, I've boiled down the problem and created a simpler gist to demonstrate the problem, in TF 2.4

```
import numpy as np
import tensorflow as tf
from tensorflow.python.ops.script_ops import _eager_py_func


def test_eager(x):
    return tf.py_function(func=np.cos, inp=[x], Tout=tf.float32)  # the problem, calls the below and passes `use_tape_cache=True`
    # return _eager_py_func(func=np.cos, inp=[x], Tout=tf.float32, use_tape_cache=False)  # calling without the cache solves the problem
    # return tf.numpy_function(func=np.cos, inp=[x], Tout=tf.float32)  # works as well



test_graph = tf.function(test_eager)
for _ in range(5000):
    test_eager(tf.random.uniform(shape=(100000,)))  # running this increases memory by ~5GB
    # test_graph(tf.random.uniform(shape=(100000,)))  # running this is fine

```

It's with the `use_tape_cache`, at least for the above. Here we can use the `numpy_func`, maybe that works as well for the data preprocessing as no gradient is needed there.

**Comment by tilakrayal** [comment]
@workingloong ,
Is this still an issue?
Could you please update TensorFlow to the latest stable version v.2.7 and let us know if you are facing the same error. Thanks!

**Comment by jonas-eschle** [comment]
I can confirm that it still exists (tested with nightlies), I've updated the snippet

**Comment by ladi-pomsar** [comment]
Can someone please finally look at this? It has been almost 3 years

**Comment by sachinprasadhs** [comment]
The memory leak seem to be addressed in the commit here https://github.com/tensorflow/tensorflow/commit/7c794ca526a61d9989ebd272155f3c76005e6e71.
Let us know if you are still facing an issue. Thanks!
